In [2]:
# reload magics
%load_ext autoreload
%autoreload 2

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datasets
from project_modules.io import load_dataset_to_df
from project_modules.classifcation import classify_MP,getXY, boruta_fs
from sklearn.ensemble import RandomForestClassifier
import multiprocessing
from sklearn.model_selection import cross_val_score

# from project_modules.utils import get_logger
# logger = get_logger("log-data-combine-split.log")
# # read the parameter file

# from project_modules.utils import read_parameters
# parms = read_parameters("/Users/david/projects/lc-project-data/project.yaml")

In [3]:
# from pathlib import Path
# input_dir = Path(parms["dir_proc_data"])
# ipath = Path(input_dir / "SYNTHETIC_dataset_dict")
# ipath

# Import Data

In [8]:
T81_tr = load_dataset_to_df("../Data/DataV1/TTS/T81_Train.arrow", verbose=True)
T81_ts = load_dataset_to_df("../Data/DataV1/TTS/T81_Test.arrow", verbose=True)
T81_h = load_dataset_to_df("../Data/DataV1/TTS/T81_Holdout.arrow", verbose=True)
T81_tr.shape, T81_ts.shape, T81_h.shape

((2842, 111), (948, 111), (948, 111))

In [9]:
T83_tr = load_dataset_to_df( "../Data/DataV1/TTS/T83_Train.arrow", verbose=True)
T83_ts = load_dataset_to_df("../Data/DataV1/TTS/T83_Test.arrow", verbose=True)
T83_h = load_dataset_to_df("../Data/DataV1/TTS/T83_Holdout.arrow", verbose=True)
T83_tr.shape, T83_ts.shape, T83_h.shape

((3315, 90), (1105, 90), (1106, 90))

In [10]:
T81_tr

sex   age  dag  MED-Analgesics  MED-Anesthetics  MED-Antibacterials  \
0     2.0  59.0    2               0                0                   0   
1     1.0  32.0    4               0                0                   0   
2     2.0  57.0    2               0                0                   0   
3     1.0  51.0    4               0                0                   0   
4     1.0  72.0    4               0                0                   0   
...   ...   ...  ...             ...              ...                 ...   
2837  1.0  65.0    2               0                0                   0   
2838  2.0  28.0    2               0                0                   0   
2839  1.0  86.0    2               0                0                   0   
2840  1.0  41.0    2               0                0                   0   
2841  1.0  61.0    2               0                0                   0   

      MED-Anticonvulsants  MED-Antidepressants   MED-Antifungals  \
0                       0                     0                0   
1                       0                     0                0   
2                       0                     0                0   
3                       0                     0                0   
4                       0                     0                0   
...                   ...                   ...              ...   
2837                    0                     0                0   
2838                    0                     0                0   
2839                    0                     0                0   
2840                    0                     0                0   
2841                    0                     0                0   

      MED-Antimigraine-agents  ...  HHX-OTHER-surgery  \
0                           0  ...                  0   
1                           0  ...                  0   
2                           0  ...                  0   
3                           0  ...                  0   
4                           0  ...                  0   
...                       ...  ...                ...   
2837                        0  ...                  0   
2838                        0  ...                  0   
2839                        0  ...                  0   
2840                        0  ...                  0   
2841                        0  ...                  0   

      HHX-OTHER-ulcerative-colitis  HHX-Obesity  HHX-Other  HHX-Pregnant  \
0                                0          0.0        0.0           0.0   
1                                0          0.0        0.0           0.0   
2                                0          0.0        1.0           0.0   
3                                0          0.0        0.0           0.0   
4                                0          0.0        0.0           0.0   
...                            ...          ...        ...           ...   
2837                             0          0.0        0.0           0.0   
2838                             0          0.0        0.0           0.0   
2839                             0          0.0        0.0           0.0   
2840                             0          1.0        1.0           0.0   
2841                             0          0.0        0.0           0.0   

      SUBST-1  SUBST-5  LC_STATUS  __index_level_0__  LC_STATUS_DAG  
0         1.0      0.0     LC_NEG               1364       LC_NEG_2  
1         0.0      0.0     LC_POS               3561       LC_POS_4  
2         0.0      0.0     LC_NEG               2409       LC_NEG_2  
3         0.0      0.0     LC_POS               3102       LC_POS_4  
4         0.0      0.0     LC_POS               3333       LC_POS_4  
...       ...      ...        ...                ...            ...  
2837      1.0      0.0     LC_NEG               2284       LC_NEG_2  
2838      0.0      0.0     LC_NEG               2651       LC_NEG_2  
2839      0.0      0.0     LC_NEG   

In [11]:
lDataFrames = [T81_tr, T81_ts, T81_h, T83_tr, T83_ts, T83_h]
lColDrop = ["__index_level_0__", "LC_STATUS_DAG"]

for df in lDataFrames:
    df.drop(df[df['LC_STATUS']=='HC'].index, inplace=True) # drop HC
    df['LC_STATUS'] = df['LC_STATUS'].apply(lambda x: 1 if x == 'LC_POS' else 0) #Convert to 0==LCNeg, 1==LCPos
    df.drop(columns=lColDrop, inplace=True) # drop unneeded columns

In [13]:
T81_tr.shape,T83_tr.shape

((2794, 109), (3267, 88))

In [21]:
lScorersBinary = [
    "accuracy",
    "balanced_accuracy",
    "roc_auc",
    "f1",
    "recall",
    "sensitivity",
    "specificity",
    "precision",
    # "average_precision",
    "NPV",
    "PPV",
    # "neg_mean_squared_error",
]

# Basic RF Classifier - Holdout sets, all Features

In [11]:

lClassifiers = ["RF", "ET", "LogR"] # different classifiers
lCV = [3,5,10] # different cross-validation splits
lTrees=[100,1000,10000] #num trees
lMD=[None] #max depth

## T83 - Holdout

In [8]:
x,y = getXY(T83_h)

In [9]:
dfResults = classify_MP(X=x, y=y, lScorers=lScorersBinary, lClassifiers=lClassifiers, lTrees=lTrees, lMD=lMD,lCV=lCV)

In [10]:
# dfResults.columns

In [11]:
lResCol = ['Title','cv']+[x for x in dfResults.columns if x.startswith("param_") or x.startswith("mean_test_")]

In [12]:
dfResults['Title']='RT83_Holdout_LCNeg_LCPos'

/tmp/ipykernel_38070/3276964138.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dfResults['Title']='RT83_Holdout_LCNeg_LCPos'


In [13]:
dfResults[lResCol]

Title  cv                 param_clf  \
0  RT83_Holdout_LCNeg_LCPos   3  RandomForestClassifier()   
1  RT83_Holdout_LCNeg_LCPos   3  RandomForestClassifier()   
2  RT83_Holdout_LCNeg_LCPos   3  RandomForestClassifier()   
3  RT83_Holdout_LCNeg_LCPos   3    ExtraTreesClassifier()   
4  RT83_Holdout_LCNeg_LCPos   3    ExtraTreesClassifier()   
5  RT83_Holdout_LCNeg_LCPos   3    ExtraTreesClassifier()   
6  RT83_Holdout_LCNeg_LCPos   3      LogisticRegression()   
0  RT83_Holdout_LCNeg_LCPos   5  RandomForestClassifier()   
1  RT83_Holdout_LCNeg_LCPos   5  RandomForestClassifier()   
2  RT83_Holdout_LCNeg_LCPos   5  RandomForestClassifier()   
3  RT83_Holdout_LCNeg_LCPos   5    ExtraTreesClassifier()   
4  RT83_Holdout_LCNeg_LCPos   5    ExtraTreesClassifier()   
5  RT83_Holdout_LCNeg_LCPos   5    ExtraTreesClassifier()   
6  RT83_Holdout_LCNeg_LCPos   5      LogisticRegression()   
0  RT83_Holdout_LCNeg_LCPos  10  RandomForestClassifier()   
1  RT83_Holdout_LCNeg_LCPos  10  RandomForestClassifier()   
2  RT83_Holdout_LCNeg_LCPos  10  RandomForestClassifier()   
3  RT83_Holdout_LCNeg_LCPos  10    ExtraTreesClassifier()   
4  RT83_Holdout_LCNeg_LCPos  10    ExtraTreesClassifier()   
5  RT83_Holdout_LCNeg_LCPos  10    ExtraTreesClassifier()   
6  RT83_Holdout_LCNeg_LCPos  10      LogisticRegression()   

  param_clf__max_depth param_clf__n_estimators param_clf__random_state  \
0                 None                     100                      42   
1                 None                    1000                      42   
2                 None                   10000                      42   
3                 None                     100                      42   
4                 None                    1000                      42   
5                 None                   10000                      42   
6                  NaN                     NaN                     NaN   
0                 None                     100                      42   
1                 None                    1000                      42   
2                 None                   10000                      42   
3                 None                     100                      42   
4                 None                    1000                      42   
5                 None                   10000                      42   
6                  NaN                     NaN                     NaN   
0                 None                     100                      42   
1                 None                    1000                      42   
2                 None                   10000                      42   
3                 None                     100                      42   
4                 None                    1000                      42   
5                 None                   10000                      42   
6                  NaN                     NaN                     NaN   

  param_clf__max_iter  mean_test_accuracy  mean_test_balanced_accuracy  \
0                 NaN            0.719008                     0.697772   
1                 NaN            0.719008                     0.697772   
2                 NaN            0.715335                     0.694520   
3                 NaN            0.707071                     0.690347   
4                 NaN            0.707989                     0.691160   
5                 NaN            0.707071                     0.690347   
6                1000            0.651974                     0.639632   
0                 NaN            0.713491                     0.692608   
1                 NaN            0.714413                     0.694630   
2                 NaN            0.715326                     0.695682   
3                 NaN            0.700655                     0.685115   
4                 NaN            0.703391                     0.688491   
5                 NaN            0.703382                     0.688970   
6             

In [14]:
dfResults.to_csv("OUTPUT/MP/05-classifiers/CA_T83_Holdout_CV3510.csv", index=False)
dfResults[lResCol].to_csv("OUTPUT/MP/05-classifiers/CA_T83_Holdout_CV3510_selCol.csv", index=False)

## T81 - Holdout

In [15]:
x,y = getXY(T81_h)

In [16]:
dfResults = classify_MP(X=x, y=y, lScorers=lScorersBinary, lClassifiers=lClassifiers, lTrees=lTrees, lMD=lMD,lCV=lCV)

In [17]:
# dfResults.columns

In [18]:
lResCol = ['Title','cv']+[x for x in dfResults.columns if x.startswith("param_") or x.startswith("mean_test_")]

In [19]:
dfResults['Title']='RT81_Holdout_LCNeg_LCPos'

/tmp/ipykernel_38070/1820217221.py:1: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  dfResults['Title']='RT81_Holdout_LCNeg_LCPos'


In [20]:
dfResults[lResCol]

Title  cv                 param_clf  \
0  RT81_Holdout_LCNeg_LCPos   3  RandomForestClassifier()   
1  RT81_Holdout_LCNeg_LCPos   3  RandomForestClassifier()   
2  RT81_Holdout_LCNeg_LCPos   3  RandomForestClassifier()   
3  RT81_Holdout_LCNeg_LCPos   3    ExtraTreesClassifier()   
4  RT81_Holdout_LCNeg_LCPos   3    ExtraTreesClassifier()   
5  RT81_Holdout_LCNeg_LCPos   3    ExtraTreesClassifier()   
6  RT81_Holdout_LCNeg_LCPos   3      LogisticRegression()   
0  RT81_Holdout_LCNeg_LCPos   5  RandomForestClassifier()   
1  RT81_Holdout_LCNeg_LCPos   5  RandomForestClassifier()   
2  RT81_Holdout_LCNeg_LCPos   5  RandomForestClassifier()   
3  RT81_Holdout_LCNeg_LCPos   5    ExtraTreesClassifier()   
4  RT81_Holdout_LCNeg_LCPos   5    ExtraTreesClassifier()   
5  RT81_Holdout_LCNeg_LCPos   5    ExtraTreesClassifier()   
6  RT81_Holdout_LCNeg_LCPos   5      LogisticRegression()   
0  RT81_Holdout_LCNeg_LCPos  10  RandomForestClassifier()   
1  RT81_Holdout_LCNeg_LCPos  10  RandomForestClassifier()   
2  RT81_Holdout_LCNeg_LCPos  10  RandomForestClassifier()   
3  RT81_Holdout_LCNeg_LCPos  10    ExtraTreesClassifier()   
4  RT81_Holdout_LCNeg_LCPos  10    ExtraTreesClassifier()   
5  RT81_Holdout_LCNeg_LCPos  10    ExtraTreesClassifier()   
6  RT81_Holdout_LCNeg_LCPos  10      LogisticRegression()   

  param_clf__max_depth param_clf__n_estimators param_clf__random_state  \
0                 None                     100                      42   
1                 None                    1000                      42   
2                 None                   10000                      42   
3                 None                     100                      42   
4                 None                    1000                      42   
5                 None                   10000                      42   
6                  NaN                     NaN                     NaN   
0                 None                     100                      42   
1                 None                    1000                      42   
2                 None                   10000                      42   
3                 None                     100                      42   
4                 None                    1000                      42   
5                 None                   10000                      42   
6                  NaN                     NaN                     NaN   
0                 None                     100                      42   
1                 None                    1000                      42   
2                 None                   10000                      42   
3                 None                     100                      42   
4                 None                    1000                      42   
5                 None                   10000                      42   
6                  NaN                     NaN                     NaN   

  param_clf__max_iter  mean_test_accuracy  mean_test_balanced_accuracy  \
0                 NaN            0.777648                     0.773656   
1                 NaN            0.772292                     0.768335   
2                 NaN            0.772295                     0.768662   
3                 NaN            0.766916                     0.762836   
4                 NaN            0.766909                     0.763345   
5                 NaN            0.765840                     0.762486   
6                1000            0.758310                     0.754640   
0                 NaN            0.784072                     0.780397   
1                 NaN            0.784095                     0.781414   
2                 NaN            0.783020                     0.780577   
3                 NaN            0.774424                     0.772224   
4                 NaN            0.783020                     0.780563   
5                 NaN            0.781945                     0.779875   
6             

In [21]:
dfResults.to_csv("OUTPUT/MP/05-classifiers/CA_T81_Holdout_CV3510.csv", index=False)
dfResults[lResCol].to_csv("OUTPUT/MP/05-classifiers/CA_T81_Holdout_CV3510_selCol.csv", index=False)

## Summary
T81 Metrics Superior to T83... likely some important columns there

Main improvmeent in sensitivity/recall

In [5]:
lResCol = [
    "Title",
    "cv",
    "param_clf",
    "param_clf__max_depth",
    "param_clf__n_estimators",
    "param_clf__random_state",
    # "param_clf__max_iter",
    "mean_test_accuracy",
    "mean_test_balanced_accuracy",
    "mean_test_roc_auc",
    "mean_test_f1",
    "mean_test_recall",
    "mean_test_sensitivity",
    "mean_test_specificity",
    "mean_test_precision",
    "mean_test_NPV",
    "mean_test_PPV",
]

In [7]:
rT81 = pd.read_csv("OUTPUT/MP/05-classifiers/CA_T81_Holdout_CV3510.csv")
rT83 = pd.read_csv("OUTPUT/MP/05-classifiers/CA_T83_Holdout_CV3510.csv")
dfR = pd.concat([rT81,rT83])[lResCol]
dfR[(dfR['cv']==5) & (dfR['param_clf']=='RandomForestClassifier()') & (dfR['param_clf__n_estimators']==1000)].T

8  \
Title                        RT81_Holdout_LCNeg_LCPos   
cv                                                  5   
param_clf                    RandomForestClassifier()   
param_clf__max_depth                              NaN   
param_clf__n_estimators                        1000.0   
param_clf__random_state                          42.0   
mean_test_accuracy                           0.784095   
mean_test_balanced_accuracy                  0.781414   
mean_test_roc_auc                            0.876196   
mean_test_f1                                   0.7619   
mean_test_recall                             0.744828   
mean_test_sensitivity                        0.744828   
mean_test_specificity                           0.818   
mean_test_precision                          0.780777   
mean_test_NPV                                0.787887   
mean_test_PPV                                0.780777   

                                                    8  
Title                        RT83_Holdout_LCNeg_LCPos  
cv                                                  5  
param_clf                    RandomForestClassifier()  
param_clf__max_depth                              NaN  
param_clf__n_estimators                        1000.0  
param_clf__random_state                          42.0  
mean_test_accuracy                           0.714413  
mean_test_balanced_accuracy                   0.69463  
mean_test_roc_auc                            0.759707  
mean_test_f1                                 0.622592  
mean_test_recall                             0.542105  
mean_test_sensitivity                        0.542105  
mean_test_specificity                        0.847154  
mean_test_precision                           0.73342  
mean_test_NPV                                0.706186  
mean_test_PPV                                 0.73342

# Boruta FS

In [105]:
## Below runs all permutations of different datasets of interest at different parameters -- saves featuresets externally
# Completed Runs:
# lMD = [3,5,7]
# lTitle = ['T81','T83','T81noDag','T83noDat']
# lDatasets=[T81_tr,T83_tr,T81_tr.drop(columns=['dag']),T83_tr.drop(columns=['dag'])]

lMD = [3,5]
lTitle = ['T81noDag','T83noDat']
lDatasets=[T81_tr.drop(columns=['dag']),T83_tr.drop(columns=['dag'])]

for i, ds in enumerate(lDatasets):
    x, y = getXY(ds)
    for md in lMD:
        print(f'------{lTitle[i]}, md={md}')
        _, _ = boruta_fs(X=x.values, y=y,feat_list=x.columns, trees=1000,ittr=500,threshold=100,top_rank=5,verbose=0,model=RandomForestClassifier(n_jobs=-1,class_weight='balanced',max_depth=md,random_state=42),fileName=f"OUTPUT/MP/05-classifiers/FS/FS_{lTitle[i]}_Boruta_T1000_itrr500_thres100_topR5_MD{md}.csv")

------T81noDag, 3
1	True	age
1	True	MED-Cardiovascular-agents
1	True	SYMPT-abdominal_pain___2
5	False	SYMPT-chest_pain___1
1	True	SYMPT-chest_pain___2
1	True	SYMPT-confusion_altered_mental_s___2
1	True	SYMPT-cough___1
1	True	SYMPT-cough___2
1	True	SYMPT-diarrhea___2
1	True	SYMPT-dizziness___1
1	True	SYMPT-dizziness___2
1	True	SYMPT-extremity_weakness_or_numb___1
1	True	SYMPT-extremity_weakness_or_numb___2
1	True	SYMPT-fatigue___1
1	True	SYMPT-fatigue___2
1	True	SYMPT-fever_38_0oc___1
1	True	SYMPT-fever_38_0oc___2
1	True	SYMPT-headache___1
1	True	SYMPT-headache___2
1	True	SYMPT-joint_pain_arthralgia___2
1	True	SYMPT-leg_swelling_edema___2
1	True	SYMPT-loss_of_appetite___1
1	True	SYMPT-loss_of_appetite___2
2	False	SYMPT-loss_of_taste_lost_of_smel___1
1	True	SYMPT-loss_of_taste_lost_of_smel___2
1	True	SYMPT-muscle_aches_myalgia___1
1	True	SYMPT-muscle_aches_myalgia___2
1	True	SYMPT-nausea_vomiting___2
1	True	SYMPT-shortness_of_breath_dyspne___1
1	True	SYMPT-shortness_of_breath_dyspne___2


# Test Boruta Produced Feature Sets

In [16]:
import glob

folder_path = 'OUTPUT/MP/05-classifiers/FS'  # Replace with the actual folder of boruta exported features

csv_files = glob.glob(folder_path + '/*Boruta*.csv')

# Get all featuresets identified and save them to dFeat
dFeat = {}
for file in csv_files:
    t = file.replace(folder_path + '/FS_','').replace('.csv','')
    dfTmp = pd.read_csv(file,names=['rank','trueFeat','feat'])
    trueFeatTmp = dfTmp[dfTmp["trueFeat"] == True]['feat'].to_list()
    topFeatTmp = dfTmp["feat"].to_list()
    dFeat[t+'_True']=trueFeatTmp
    dFeat[t+'_Top']=topFeatTmp
    print(t)


T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD7
T81_Boruta_T1000_itrr200_thres100_topR5_MD7
T81noDag_Boruta_T1000_itrr200_thres100_topR5_MD7
T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD3
T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD5
T81_Boruta_T1000_itrr200_thres100_topR5_MD3
T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD5
T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD3
T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD3
T83_Boruta_T1000_itrr200_thres100_topR5_MD7
T83_Boruta_T1000_itrr200_thres100_topR5_MD3


In [17]:
# Function runs a simple test of ideal parameteres T=1000, RF, MD=None, cv=5
def test_reduced_features(df,featRedu,title):
    x,y = getXY(df)
    x = x[featRedu]
    dfRTmp = classify_MP(
        X=x,
        y=y,
        lScorers=lScorersBinary,
        lClassifiers=["RF"],
        lTrees=[1000],
        lMD=[None],
        lCV=[5],
    )
    dfRTmp["Title"] = f"Test_{title}_LCN_LCP"
    return dfRTmp
def test_reduced_features_MP(inputTuple):
    df,featRedu,title = inputTuple
    x,y = getXY(df)
    x = x[featRedu]
    dfRTmp = classify_MP(
        X=x,
        y=y,
        lScorers=lScorersBinary,
        lClassifiers=["RF"],
        lTrees=[1000],
        lMD=[None],
        lCV=[5],
    )
    dfRTmp["Title"] = f"Test_{title}_LCN_LCP"
    return dfRTmp

In [18]:
from tqdm.contrib.concurrent import process_map

In [19]:
lResCol = [
    "Title",
    "cv",
    "param_clf",
    "param_clf__max_depth",
    "param_clf__n_estimators",
    "param_clf__random_state",
    # "param_clf__max_iter",
    "mean_test_accuracy",
    "mean_test_balanced_accuracy",
    "mean_test_roc_auc",
    "mean_test_f1",
    "mean_test_recall",
    "mean_test_sensitivity",
    "mean_test_specificity",
    "mean_test_precision",
    "mean_test_NPV",
    "mean_test_PPV",
]

In [22]:
# Go through all featuresets identified and determine metrics
lTestResults = []
lArgs = []
for k,v in dFeat.items():
    if 'T81' in k:
        dataset=T81_ts
        # print('T81',v,k)
    elif 'T83' in k:
        dataset=T83_ts
        # print('T83',v,k)
    else:
        raise Exception("Unknown dataset")
    lArgs.append((dataset,v,k))

with multiprocessing.Pool(20) as pool:
    lTestResults = pool.map(test_reduced_features_MP, lArgs)
    # lTestResults.append(test_reduced_features(dataset,v,k))


/home/mtrx/miniconda3/envs/clust/lib/python3.11/site-packages/joblib/parallel.py:1332: UserWarning: Loky-backed parallel loops cannot be called in a multiprocessing, setting n_jobs=1
  n_jobs = self._backend.configure(n_jobs=self.n_jobs, parallel=self,
/home/mtrx/miniconda3/envs/clust/lib/python3.11/site-packages/joblib/parallel.py:1332: UserWarning: Loky-backed parallel loops cannot be called in a multiprocessing, setting n_jobs=1
  n_jobs = self._backend.configure(n_jobs=self.n_jobs, parallel=self,
/home/mtrx/miniconda3/envs/clust/lib/python3.11/site-packages/joblib/parallel.py:1332: UserWarning: Loky-backed parallel loops cannot be called in a multiprocessing, setting n_jobs=1
  n_jobs = self._backend.configure(n_jobs=self.n_jobs, parallel=self,
/home/mtrx/miniconda3/envs/clust/lib/python3.11/site-packages/joblib/parallel.py:1332: UserWarning: Loky-backed parallel loops cannot be called in a multiprocessing, setting n_jobs=1
  n_jobs = self._backend.configure(n_jobs=self.n_jobs, par

In [28]:
pd.set_option("display.max_colwidth", None)
dfTestResults = pd.concat(lTestResults)[lResCol].sort_values(
    "mean_test_accuracy", ascending=False
)
dfTestResults

Title  cv  \
0       Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP   5   
0        Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_Top_LCN_LCP   5   
0  Test_T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD3_True_LCN_LCP   5   
0       Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD3_True_LCN_LCP   5   
0        Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD3_Top_LCN_LCP   5   
0   Test_T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD3_Top_LCN_LCP   5   
0        Test_T83_Boruta_T1000_itrr200_thres100_topR5_MD3_Top_LCN_LCP   5   
0       Test_T83_Boruta_T1000_itrr200_thres100_topR5_MD3_True_LCN_LCP   5   
0   Test_T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD5_Top_LCN_LCP   5   
0        Test_T83_Boruta_T1000_itrr200_thres100_topR5_MD7_Top_LCN_LCP   5   
0  Test_T81noDag_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP   5   
0   Test_T81noDag_Boruta_T1000_itrr200_thres100_topR5_MD7_Top_LCN_LCP   5   
0  Test_T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD5_True_LCN_LCP   5   
0       Test_T83_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP   5   
0   Test_T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD3_Top_LCN_LCP   5   
0   Test_T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD3_Top_LCN_LCP   5   
0   Test_T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD5_Top_LCN_LCP   5   
0  Test_T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD3_True_LCN_LCP   5   
0  Test_T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD3_True_LCN_LCP   5   
0  Test_T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD5_True_LCN_LCP   5   
0   Test_T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD7_Top_LCN_LCP   5   
0  Test_T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP   5   

                  param_clf param_clf__max_depth param_clf__n_estimators  \
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   
0  RandomForestClassifier()                 None                    1000   

  param_clf__random_state  mean_test_accuracy  mean_test_balanced_accuracy  \
0                      42            0.770416                     0.766269   
0                      42            0.759692                     0.758607   
0                      42            0.729630                     0.725180   
0                      42            0.729630                     0.725180   
0                      42            0.729625                     0.725342   
0                      42            0.729625                     0.725342 

In [31]:
dfTestResults[['Title']+[x for x in dfTestResults.columns if x.startswith("mean_test_")]]

Title  \
0       Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP   
0        Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_Top_LCN_LCP   
0  Test_T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD3_True_LCN_LCP   
0       Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD3_True_LCN_LCP   
0        Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD3_Top_LCN_LCP   
0   Test_T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD3_Top_LCN_LCP   
0        Test_T83_Boruta_T1000_itrr200_thres100_topR5_MD3_Top_LCN_LCP   
0       Test_T83_Boruta_T1000_itrr200_thres100_topR5_MD3_True_LCN_LCP   
0   Test_T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD5_Top_LCN_LCP   
0        Test_T83_Boruta_T1000_itrr200_thres100_topR5_MD7_Top_LCN_LCP   
0  Test_T81noDag_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP   
0   Test_T81noDag_Boruta_T1000_itrr200_thres100_topR5_MD7_Top_LCN_LCP   
0  Test_T81noDag_Boruta_T1000_itrr500_thres100_topR5_MD5_True_LCN_LCP   
0       Test_T83_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP   
0   Test_T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD3_Top_LCN_LCP   
0   Test_T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD3_Top_LCN_LCP   
0   Test_T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD5_Top_LCN_LCP   
0  Test_T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD3_True_LCN_LCP   
0  Test_T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD3_True_LCN_LCP   
0  Test_T83noDat_Boruta_T1000_itrr500_thres100_topR5_MD5_True_LCN_LCP   
0   Test_T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD7_Top_LCN_LCP   
0  Test_T83noDag_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP   

   mean_test_accuracy  mean_test_balanced_accuracy  mean_test_roc_auc  \
0            0.770416                     0.766269           0.817495   
0            0.759692                     0.758607           0.843298   
0            0.729630                     0.725180           0.817960   
0            0.729630                     0.725180           0.817960   
0            0.729625                     0.725342           0.816219   
0            0.729625                     0.725342           0.816219   
0            0.726343                     0.713939           0.774948   
0            0.714413                     0.701439           0.763923   
0            0.709269                     0.706220           0.803956   
0            0.702448                     0.689859           0.757975   
0            0.701719                     0.699435           0.778600   
0            0.700684                     0.698170           0.790902   
0            0.700684                     0.697741           0.793134   
0            0.689617                     0.679448           0.760671   
0            0.665776                     0.657358           0.709094   
0            0.665776                     0.657358           0.709094   
0            0.632710                     0.623024           0.679874   
0            0.625350                     0.617452           0.671342   
0            0.625350                     0.617452           0.671342   
0            0.608832                     0.604010           0.659393   
0            0.607881                     0.602930           0.638811   
0            0.575775                     0.556583           0.574495   

   mean_test_f1  mean_test_recall  mean_test_sensitivity  \
0      0.741037          0.708982               0.708982   
0      0.741805          0.743598               0.743598   
0      0.694367          0.662764               0.662764   
0      0.694367          0.662764               0.662764   
0      0.695202          0.665170               0.665170   
0      0.695202          0.665170               0.665170   
0      0.662521          0.616842               0.616842   
0      0.646121          0.600000               0.600000   
0      0.678488          0.662924               0.662924   
0      0.634234          0.591579               0.591579   
0      0.675087          0.667415           

In [29]:
# dfTestResults.to_csv("OUTPUT/MP/05-classifiers/FS/FS_TestDatasetResults_T81_T83.csv", index=False)

## Summary - Best set so far seems to be Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP

# Holdoutset - validation testing - Boruta Tested Best Feature

In [ ]:
lScorersBinary = [
    "accuracy",
    "balanced_accuracy",
    "roc_auc",
    "f1",
    "recall",
    "sensitivity",
    "specificity",
    "precision",
    # "average_precision",
    "NPV",
    "PPV",
    # "neg_mean_squared_error",
]
lClassifiers = ["RF", "ET", "LogR"] # different classifiers
lCV = [3,5,10] # different cross-validation splits
lTrees=[100,1000,10000] #num trees
lMD=[None] #max depth


In [33]:
selFeat = "Test_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True_LCN_LCP".replace(
    "_LCN_LCP", ""
).replace("Test_", "")
lSelFeat = dFeat[selFeat]
x,y = getXY(T81_h)
x=x[lSelFeat]

In [47]:
dfResults = classify_MP(X=x, y=y, lScorers=lScorersBinary, lClassifiers=lClassifiers, lTrees=lTrees, lMD=lMD,lCV=lCV)

In [ ]:
# dfResults.columns

In [55]:
dfResults['Title']='R81_Holdout_'+selFeat

In [56]:
lResCol = ['Title','cv']+[x for x in dfResults.columns if x.startswith("param_") or x.startswith("mean_test_")]

In [71]:
dfResults[lResCol] # all results

Title  cv  \
0  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   3   
1  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   3   
2  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   3   
3  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   3   
4  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   3   
5  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   3   
6  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   3   
0  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   5   
1  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   5   
2  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   5   
3  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   5   
4  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   5   
5  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   5   
6  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   5   
0  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True  10   
1  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True  10   
2  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True  10   
3  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True  10   
4  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True  10   
5  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True  10   
6  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True  10   

                  param_clf param_clf__max_depth param_clf__n_estimators  \
0  RandomForestClassifier()                 None                     100   
1  RandomForestClassifier()                 None                    1000   
2  RandomForestClassifier()                 None                   10000   
3    ExtraTreesClassifier()                 None                     100   
4    ExtraTreesClassifier()                 None                    1000   
5    ExtraTreesClassifier()                 None                   10000   
6      LogisticRegression()                  NaN                     NaN   
0  RandomForestClassifier()                 None                     100   
1  RandomForestClassifier()                 None                    1000   
2  RandomForestClassifier()                 None                   10000   
3    ExtraTreesClassifier()                 None                     100   
4    ExtraTreesClassifier()                 None                    1000   
5    ExtraTreesClassifier()                 None                   10000   
6      LogisticRegression()                  NaN                     NaN   
0  RandomForestClassifier()                 None                     100   
1  RandomForestClassifier()                 None                    1000   
2  RandomForestClassifier()                 None                   10000   
3    ExtraTreesClassifier()                 None                     100   
4    ExtraTreesClassifier()                 None                    1000   
5    ExtraTreesClassifier()                 None                   10000   
6      LogisticRegression()                  NaN                     NaN   

  param_clf__random_state param_clf__max_iter  mean_test_accuracy  \
0                      42                 NaN            0.702455   
1                      42                 NaN            0.701390   
2                      42                 NaN            0.705681   
3                      42                 NaN            0.692784   
4                      42                 NaN            0.694935   
5                      42                 NaN            0.696010   
6                     NaN                1000            0.770128   
0                      42                 NaN            0.716405   
1                      42                 NaN            0.725007   
2                      42                 NaN            0.723943   
3                      42                 NaN            0.7196

In [39]:
dfResults[
    (dfResults["cv"] == 5)
    # & (dfResults["param_clf"] == "RandomForestClassifier()")
    & (dfResults["param_clf__n_estimators"] == 1000)
][lResCol].T

8   \
Title                        R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   
cv                                                                                      5   
param_clf                                                        RandomForestClassifier()   
param_clf__max_depth                                                                  NaN   
param_clf__n_estimators                                                            1000.0   
param_clf__random_state                                                              42.0   
mean_test_accuracy                                                               0.725007   
mean_test_balanced_accuracy                                                      0.721461   
mean_test_roc_auc                                                                0.797399   
mean_test_f1                                                                     0.692943   
mean_test_recall                                                                 0.672922   
mean_test_sensitivity                                                            0.672922   
mean_test_specificity                                                                0.77   
mean_test_precision                                                              0.717786   
mean_test_NPV                                                                    0.733738   
mean_test_PPV                                                                    0.717786   

                                                                                       11  
Title                        R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True  
cv                                                                                      5  
param_clf                                                          ExtraTreesClassifier()  
param_clf__max_depth                                                                  NaN  
param_clf__n_estimators                                                            1000.0  
param_clf__random_state                                                              42.0  
mean_test_accuracy                                                               0.715336  
mean_test_balanced_accuracy                                                      0.709272  
mean_test_roc_auc                                                                0.751675  
mean_test_f1                                                                     0.669664  
mean_test_recall                                                                 0.626544  
mean_test_sensitivity                                                            0.626544  
mean_test_specificity                                                               0.792  
mean_test_precision                                                              0.724707  
mean_test_NPV                                                                    0.712853  
mean_test_PPV                                                                    0.724707

In [38]:
dfResults[(dfResults['cv']==5) & (dfResults['param_clf']=='RandomForestClassifier()') & (dfResults['param_clf__n_estimators']==1000)][lResCol]

Title  cv  \
8  R81_Holdout_T81_Boruta_T1000_itrr200_thres100_topR5_MD7_True   5   

                  param_clf  param_clf__max_depth  param_clf__n_estimators  \
8  RandomForestClassifier()                   NaN                   1000.0   

   param_clf__random_state  mean_test_accuracy  mean_test_balanced_accuracy  \
8                     42.0            0.725007                     0.721461   

   mean_test_roc_auc  mean_test_f1  mean_test_recall  mean_test_sensitivity  \
8           0.797399      0.692943          0.672922               0.672922   

   mean_test_specificity  mean_test_precision  mean_test_NPV  mean_test_PPV  
8                   0.77             0.717786       0.733738       0.717786

In [79]:
dfResults.to_csv(
    f"OUTPUT/MP/05-classifiers/Validation/CA_{'R81_Holdout_'+selFeat}.csv", index=False
)
dfResults[lResCol].to_csv(f"OUTPUT/MP/05-classifiers/Validation/CA_{'R81_Holdout_'+selFeat}_selCol.csv", index=False)

In [37]:
dfResults = pd.read_csv(
    f"OUTPUT/MP/05-classifiers/Validation/CA_{'R81_Holdout_'+selFeat}.csv"
)

### Feature Importance of Best Set


In [61]:
clf = RandomForestClassifier(n_estimators=1000,max_depth=None,random_state=42)
clf.fit(x,y)
feature_importances = clf.feature_importances_
feature_importances = pd.Series(feature_importances, index=x.columns).sort_values(
    ascending=False
)

In [62]:
feature_importances

age                                     0.419838
dag                                     0.235372
SYMPT-fatigue___2                       0.084360
SYMPT-dizziness___2                     0.079019
SYMPT-muscle_aches_myalgia___2          0.074982
SYMPT-shortness_of_breath_dyspne___2    0.059716
SYMPT-joint_pain_arthralgia___2         0.046712
dtype: float64

In [76]:
feature_importances.to_csv(f"OUTPUT/MP/05-classifiers/Validation/FR_{'R81_Holdout_'+selFeat}.csv", index=False)

age                                     0.419838
dag                                     0.235372
SYMPT-fatigue___2                       0.084360
SYMPT-dizziness___2                     0.079019
SYMPT-muscle_aches_myalgia___2          0.074982
SYMPT-shortness_of_breath_dyspne___2    0.059716
SYMPT-joint_pain_arthralgia___2         0.046712

In [65]:
clf = RandomForestClassifier(n_estimators=1000, max_depth=None, random_state=42)
cross_val_score(clf, x, y, cv=5, scoring="accuracy").mean()

0.7250071876257836